File: examples/DesignPatterns/FactoryMethod/java/src/main/java/bead/ex/FactoryMethod/reducers/Reducer.java
package bead.ex.FactoryMethod.reducers;

import bead.ex.FactoryMethod.dispatchers.Dispatcher;
import bead.ex.FactoryMethod.models.Action;
import bead.ex.FactoryMethod.models.ActionResult;

public abstract class Reducer<T> {
    public abstract Dispatcher<T> createDispatcher();

    public ActionResult<T> reduce(Action<T> action) {
        Dispatcher<T> dispatcher = createDispatcher();
        return dispatcher.dispatch(action);
    }
}

File: examples/DesignPatterns/FactoryMethod/typescript/src/reducers/Reducer.ts
import { Dispatcher } from '../dispatchers';
import { Action, ActionResult } from '../models';

export abstract class Reducer<T> {
  public abstract createDispatcher(): Dispatcher<T>;

  public reduce(action: Action<T>): ActionResult<T> {
    const dispatcher = this.createDispatcher();
    return dispatcher.dispatch(action);
  }
}
